pallets/farming/src/lib.rs

```diff
@@ -0,0 +1,927 @@
+        ) -> DispatchResult {
+            let exchanger = ensure_signed(origin)?;
+
+            let pool_info = Self::pool_infos(pid).ok_or(Error::<T>::PoolDoesNotExist)?;
+            ensure!(
+                PoolState::state_valid(Action::Withdraw, pool_info.state),
+                Error::<T>::InvalidPoolState
+            );
+
+            let share_info = Self::shares_and_withdrawn_rewards(pid, &exchanger)
+                .ok_or(Error::<T>::ShareInfoNotExists)?;
+            ensure!(
+                share_info.withdraw_list.len() < pool_info.withdraw_limit_count.into(),
+                Error::<T>::WithdrawLimitCountExceeded
+            );
+
+            Self::remove_share(&exchanger, pid, remove_value, pool_info.withdraw_limit_time)?;
+
+            Self::deposit_event(Event::Withdrawn {
+                who: exchanger,
+                pid,
+                remove_value,
+            });
+            Ok(())
+        }
+
+        /// User can claim rewards and also unStake.
+        #[pallet::call_index(4)]
+        #[pallet::weight(T::WeightInfo::claim())]
+        pub fn claim(origin: OriginFor<T>, pid: PoolId) -> DispatchResult {
+            let exchanger = ensure_signed(origin)?;
+
+            let pool_info = Self::pool_infos(pid).ok_or(Error::<T>::PoolDoesNotExist)?;
+            ensure!(
+                PoolState::state_valid(Action::Claim, pool_info.state),
+                Error::<T>::InvalidPoolState
+            );
+
+            let current_block_number: BlockNumberFor<T> = frame_system::Pallet::<T>::block_number();
+            let share_info = Self::shares_and_withdrawn_rewards(pid, &exchanger)
+                .ok_or(Error::<T>::ShareInfoNotExists)?;
+            ensure!(
+                share_info.claim_last_block + pool_info.claim_limit_time <= current_block_number,
+                Error::<T>::CanNotClaim
+            );
+
+            Self::claim_rewards(&exchanger, pid)?;
+            if let Some(ref gid) = pool_info.gauge {
+                Self::gauge_claim_inner(&exchanger, *gid)?;
+            }
+            Self::process_withdraw_list(&exchanger, pid, &pool_info, true)?;
+
+            Self::deposit_event(Event::Claimed {
+                who: exchanger,
+                pid,
+            });
+            Ok(())
+        }
+
+        /// User can unStake token from farm pool.
+        #[pallet::call_index(5)]
+        #[pallet::weight(T::WeightInfo::claim())]
+        pub fn withdraw_claim(origin: OriginFor<T>, pid: PoolId) -> DispatchResult {
+            let exchanger = ensure_signed(origin)?;
+
+            let pool_info = Self::pool_infos(pid).ok_or(Error::<T>::PoolDoesNotExist)?;
+            Self::process_withdraw_list(&exchanger, pid, &pool_info, false)?;
+
+            Self::deposit_event(Event::WithdrawClaimed {
+                who: exchanger,
+                pid,
+            });
+            Ok(())
+        }
+
+        #[pallet::call_index(6)]
+        #[pallet::weight(0)]
```

Comment:
still dont get why these weights are 0

Reply:
admin operation can use weight=0. This's original code, but we can benchmark it and don't allow free exectution.

Reply:
we could leave it as a separate issue, but i'd give it some non 0 value at least.

Reply:
make sense, change to 1000.